File: java/src/PegGameUtils.java
/**
 * Utility class with some basic methods used across the entire program.
 */
public class PegGameUtils {
    // This table contains the total number of pegs up to a certain row.
    public static final int[] TOTAL_PEGS_TABLE = {
            0, 1, 3, 6, 10, 15, 21, 28, 36, 45, 55, 66, 78, 91, 105, 120, 136, 153
    };

    /**
     * Returns the corresponding row for the given peg.
     */
    public static int getRow(int pegNumber) {
        int i;
        // Loop until the number of pegs becomes greater.
        for (i = 0 ; TOTAL_PEGS_TABLE[i] <= pegNumber ; i++);
        return i - 1;
    }

    /**
     * Returns the displacement from the beginning of the row.
     */
    public static int getDisplacement(int pegNumber) {
        return pegNumber - TOTAL_PEGS_TABLE[getRow(pegNumber)];
    }

    /**
     * Returns the peg number for the given row and displacement.
     */
    public static int getPegNumber(int row, int displacement) {
        if (row < 0 || row >= TOTAL_PEGS_TABLE.length || displacement < 0 ||
                displacement > row) {
            return -1;
        }
        return TOTAL_PEGS_TABLE[row] + displacement;
    }
}

File: go/src/util/PegGameUtils.go
package util

var TOTAL_PEGS_TABLE = []int {
    0, 1, 3, 6, 10, 15, 21, 28, 36, 45, 55, 66, 78, 91, 105, 120, 136, 153,
}

var MAX_ROWS = 12

var TOTAL_PEGS = 0

var NUMBER_OF_ROWS = 0

/**
 * Sets the total number of pegs for the current test.
 */
func SetTotalPegs(pegs int) {
    TOTAL_PEGS = pegs
}

/**
 * Gets the total number of pegs for the current test.
 */
func GetTotalPegs() int {
    return TOTAL_PEGS
}

/**
 * Sets the number of rows for the current test.
 */
func SetNumberOfRows(rows int) {
    NUMBER_OF_ROWS = rows
}

/**
 * Gets the number of rows for the current test.
 */
func GetNumberOfRows() int {
    return NUMBER_OF_ROWS
}

/**
 * Returns the row for the peg number provided.
 */
func GetRow(pegNumber int) int {
    i := 0
    for j := 0 ; TOTAL_PEGS_TABLE[j] <= pegNumber ; j++ {
        i = j
    }
    return i
}

/**
 * Returns the displacement for the given peg number.
 */
func GetDisplacement(pegNumber int) int {
    if (GetRow(pegNumber) < 0) {
        return -1
    }
    return pegNumber - TOTAL_PEGS_TABLE[GetRow(pegNumber)]
}

func GetPegNumber(row int, displacement int) int {
    if (row < 0 || row >= MAX_ROWS || displacement < 0 || displacement > row) {
        return -1
    }
    return TOTAL_PEGS_TABLE[row] + displacement
}
